proposal.md

```diff
@@ -1,233 +1,372 @@
+- `add(index, element)`, `removeAt(index)`, `remove(element)` - O(N).
+- `addAll(elements)` - O(N).
+- `addAll(index, elements)`, `removeAll(elements)`, `removeAll(predicate)` - O(N M), optimizable to O(N+M), where M is the number of elements to be inserted/removed.
+- Iterating elements - O(N).
 
-    // or
+To optimize frequently used `add(element)` and `removeAt(size - 1)` operations rightmost leaf is referenced directly from the persistent list instance.
+This allows to avoid path-copying and gives O(1) time complexity for these two operations.
 
-    class Query(parameters: List<Parameter>) {
-        // creates an immutable list or does nothing if it's already immutable
-        val parameters = parameters.toImmutable()
-    }
-    ```
+Small persistent lists, with up to 32 elements, are backed by arrays of corresponding size.
 
-* Sharing a single immutable collection between multiple threads. Since the collection
-cannot be changed neither by producer, nor by any consumer, you'll never get a `ConcurrentModificationException`
-during iteration, or any other symptoms of corrupted state.
+#### Persistent unordered hash set
 
-* Having multiple snapshots of a collection sharing their element storage.
+It's backed by a hash array mapped trie (a.k.a HAMT). Every node has up to 32 children or elements.
 
-## Alternatives
+Time complexity of operations:
+- `add(element)`, `remove(element)`, `contains(element)` - O(log<sub>32</sub>N) in general, but strongly depends on hash codes of the stored elements.
+- `addAll(elements)`, `removeAll(elements)`, `removeAll(predicate)`, `containsAll(elements)` - O(M log<sub>32</sub>N) in general.
+- Iterating elements - O(N).
 
-* Create a mutable collection, but expose it only as its read-only interface.
-  This pattern is mostly employed in the standard library collection operations.
+#### Persistent unordered hash map
 
-  * Pro: mutable collection implementations are generally more memory efficient,
-  require less allocations during and after their populating.
-  * Con: collection can be cast back to mutable (e.g. involving platform types),
-  and then unexpectedly mutated.
-  * Con: different snapshots of a collection do not share any element storage.
-  * Con: defensive copying of the resulting collection is still required later
-  in a chain of operations.
+It's backed by a compressed hash-array mapped prefix-tree (a.k.a CHAMP). Every node has up to 32 children or entries.
+
+Time complexity of operations:
+- `get(key)`, `put(key, value)`, `remove(key)`, `remove(key, value)`, `containsKey(key)` - O(log<sub>32</sub>N) in average, but strongly depends on hash codes of the stored keys.
+- `containsValue(value)` - O(N).
+- `putAll(map)` - O(M log<sub>32</sub>N), where M is the number of elements added.
+- Iterating `keys`, `values`, `entries` - O(N).
+
+#### Default persistent ordered hash set
```

Comment:
I believe it'd be better to describe ordered map first and then explain ordered set as a simplification of ordered map.

Reply:
Done.